File: leetcode/leetcode-198.py
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 0:
            return 0
        if n == 1:
            return nums[0]
        if n == 2:
            return max(nums[0], nums[1])
        dp = [0]*n
        ans = 0
        dp[0] = nums[0]
        dp[1] = nums[1]
        dp[2] = nums[0]+nums[2]
        ans = max(dp[0], dp[1])
        ans = max(ans, dp[2])
        for i in range(3,n):
            dp[i] = max(dp[i-2], dp[i-3])+nums[i]
            ans = max(ans, dp[i])
        return ans

if __name__ == "__main__":
    so = Solution()
    print(so.rob([2,7,9,3,1]))

File: leetcode/leetcode-198.cpp
/*
 *@File Name          :leetcode-198.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/7/28 9:24
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "vector"
using namespace std;

class Solution {
 public:
  int rob(vector<int>& nums) {
    int         n = nums.size();
    vector<int> dp(n, 0);
    dp[0] = nums[0];
    if (n == 1) return dp[n - 1];
    dp[1] = max(dp[0], nums[1]);

    for (int i = 2; i < n; i++) dp[i] = max(dp[i - 1], dp[i - 2] + nums[i]);
    return dp[n - 1];
  }
};